### PR TITLE
VS 16.5 Issues

### PR DESCRIPTION
**Pull request by kanstantsin-chernik**
Problem:
- VS 16.5 broke build

Solution:
- Employ compiler extension

Issue: #462 

Reviewers:
@krzysztof-jusiak

### CODE REVIEW

File: example/pool_provider.cpp [hunk 1]
```diff
@@ -87,7 +87,7 @@ struct pool_provider {
            ,
            TArgs&&... args) const {
     auto memory = pool_allocator::allocate<T>();
-    return std::unique_ptr<T, pool_deleter>{new (memory) T(std::forward<TArgs>(args)...)};
+    return new (memory) T(std::forward<TArgs>(args)...);
```
Comment (kanstantsin-chernik) [inline example/pool_provider.cpp:90]: This leads to an error below. Seems like provider is not supposed to return SP but rather bare pointer

```
1>pool_provider.cpp
1>C:\Projects\di\include\boost/di.hpp(2461,1): error C2440: 'return': cannot convert from 'TWrapper' to 'std::unique_ptr<interface,pool_deleter>'
1>        with
1>        [
1>            TWrapper=wrapper_t
1>        ]
1>C:\Projects\di\include\boost/di.hpp(2461,1): message : No constructor could take the source type, or constructor overload resolution was ambiguous
1>C:\Projects\di\include\boost/di.hpp(2461): message : while compiling class template member function 'boost::di::v1_1_0::core::successful::wrapper<create_t,wrapper_t>::operator T(void) noexcept'
1>        with
1>        [
1>            T=create_t
1>        ]
1>C:\Projects\di\include\boost/di.hpp(2238): message : see reference to function template instantiation 'boost::di::v1_1_0::core::successful::wrapper<create_t,wrapper_t>::operator T(void) noexcept' being compiled
1>        with
1>        [
1>            T=create_t
1>        ]
1>C:\Projects\di\include\boost/di.hpp(2959): message : see reference to class template instantiation 'boost::di::v1_1_0::core::successful::wrapper<create_t,wrapper_t>' being compiled
1>C:\Projects\di\include\boost/di.hpp(2883): message : see reference to function template instantiation 'boost::di::v1_1_0::core::successful::wrapper<create_t,wrapper_t> boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,T,TName,TPriority,TCtor>>::create_successful_impl__<TIsRoot,std::unique_ptr<interface,pool_deleter>,boost::di::v1_1_0::no_name>(void) const' being compiled
1>        with
1>        [
1>            TConfig=config,
1>            TScope=boost::di::v1_1_0::scopes::deduce,
1>            TExpected=interface,
1>            T=implementation,
1>            TName=boost::di::v1_1_0::no_name,
1>            TPriority=void,
1>            TCtor=boost::di::v1_1_0::core::none,
1>            TIsRoot=boost::di::v1_1_0::aux::false_type
1>        ]
1>C:\Projects\di\include\boost/di.hpp(2238): message : see reference to function template instantiation 'boost::di::v1_1_0::core::successful::wrapper<create_t,wrapper_t> boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,T,TName,TPriority,TCtor>>::create_successful_impl<boost::di::v1_1_0::aux::false_type,std::unique_ptr<interface,pool_deleter>>(const boost::di::v1_1_0::aux::type<std::unique_ptr<interface,pool_deleter>> &) const' being compiled
1>        with
1>        [
1>            TConfig=config,
1>            TScope=boost::di::v1_1_0::scopes::deduce,
1>            TExpected=interface,
1>            T=implementation,
1>            TName=boost::di::v1_1_0::no_name,
1>            TPriority=void,
1>            TCtor=boost::di::v1_1_0::core::none
1>        ]
1>C:\Projects\di\example\pool_provider.cpp(90): message : see reference to function template instantiation 'boost::di::v1_1_0::core::successful::any_type<TParent,boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,T,TName,TPriority,TCtor>>>::operator std::unique_ptr<interface,pool_deleter>(void)<std::unique_ptr<interface,pool_deleter>>' being compiled
1>        with
1>        [
1>            TParent=example,
1>            TConfig=config,
1>            TScope=boost::di::v1_1_0::scopes::deduce,
1>            TExpected=interface,
1>            T=implementation,
1>            TName=boost::di::v1_1_0::no_name,
1>            TPriority=void,
1>            TCtor=boost::di::v1_1_0::core::none
1>        ]
1>C:\Projects\di\example\pool_provider.cpp(90): message : see reference to function template instantiation 'boost::di::v1_1_0::core::successful::any_type<TParent,boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,T,TName,TPriority,TCtor>>>::operator std::unique_ptr<interface,pool_deleter>(void)<std::unique_ptr<interface,pool_deleter>>' being compiled
1>        with
1>        [
1>            TParent=example,
1>            TConfig=config,
1>            TScope=boost::di::v1_1_0::scopes::deduce,
1>            TExpected=interface,
1>            T=implementation,
1>            TName=boost::di::v1_1_0::no_name,
1>            TPriority=void,
1>            TCtor=boost::di::v1_1_0::core::none
1>        ]
1>C:\Projects\di\include\boost/di.hpp(2449): message : see reference to function template instantiation 'std::unique_ptr<T,pool_deleter> pool_provider::get<T,TInitialization,TMemory,boost::di::v1_1_0::core::successful::any_type<TParent,boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,implementation,TName,TPriority,TCtor>>>,boost::di::v1_1_0::core::successful::any_type<TParent,boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,implementation,TName,TPriority,TCtor>>>,boost::di::v1_1_0::core::successful::any_type<TParent,boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,implementation,TName,TPriority,TCtor>>>>(const TInitialization &,const TMemory &,boost::di::v1_1_0::core::successful::any_type<TParent,boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,implementation,TName,TPriority,TCtor>>> &&,boost::di::v1_1_0::core::successful::any_type<TParent,boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,implementation,TName,TPriority,TCtor>>> &&,boost::di::v1_1_0::core::successful::any_type<TParent,boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,implementation,TName,TPriority,TCtor>>> &&) const' being compiled
1>        with
1>        [
1>            T=example,
1>            TInitialization=boost::di::v1_1_0::type_traits::direct,
1>            TMemory=memory,
1>            TParent=example,
1>            TConfig=config,
1>            TScope=boost::di::v1_1_0::scopes::deduce,
1>            TExpected=interface,
1>            TName=boost::di::v1_1_0::no_name,
1>            TPriority=void,
1>            TCtor=boost::di::v1_1_0::core::none
1>        ]
1>C:\Projects\di\include\boost/di.hpp(1927): message : see reference to function template instantiation 'std::unique_ptr<T,pool_deleter> boost::di::v1_1_0::core::successful::provider<ctor_t,boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,implementation,TName,TPriority,TCtor>>>::get<memory>(const TMemory &) const' being compiled
1>        with
1>        [
1>            T=example,
1>            TConfig=config,
1>            TScope=boost::di::v1_1_0::scopes::deduce,
1>            TExpected=interface,
1>            TName=boost::di::v1_1_0::no_name,
1>            TPriority=void,
1>            TCtor=boost::di::v1_1_0::core::none,
1>            TMemory=memory
1>        ]
1>C:\Projects\di\include\boost/di.hpp(1156): message : see reference to function template instantiation 'wrapper boost::di::v1_1_0::scopes::unique::scope<TExpected,TGiven>::create<T,TName,TProvider>(const TProvider &) const' being compiled
1>        with
1>        [
1>            TExpected=example,
1>            TGiven=example,
1>            T=example,
1>            TName=boost::di::v1_1_0::no_name,
1>            TProvider=provider_t
1>        ]
1>C:\Projects\di\include\boost/di.hpp(2957): message : see reference to function template instantiation 'wrapper boost::di::v1_1_0::scopes::deduce::scope<TExpected,TGiven>::create<T,TName,provider_t>(const TProvider &)' being compiled
1>        with
1>        [
1>            TExpected=example,
1>            TGiven=example,
1>            T=example,
1>            TName=boost::di::v1_1_0::no_name,
1>            TProvider=provider_t
1>        ]
1>C:\Projects\di\include\boost/di.hpp(2883): message : see reference to function template instantiation 'boost::di::v1_1_0::core::successful::wrapper<create_t,wrapper_t> boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,T,TName,TPriority,TCtor>>::create_successful_impl__<TIsRoot,example,boost::di::v1_1_0::no_name>(void) const' being compiled
1>        with
1>        [
1>            TConfig=config,
1>            TScope=boost::di::v1_1_0::scopes::deduce,
1>            TExpected=interface,
1>            T=implementation,
1>            TName=boost::di::v1_1_0::no_name,
1>            TPriority=void,
1>            TCtor=boost::di::v1_1_0::core::none,
1>            TIsRoot=boost::di::v1_1_0::aux::true_type
1>        ]
1>C:\Projects\di\include\boost/di.hpp(2776): message : see reference to function template instantiation 'boost::di::v1_1_0::core::successful::wrapper<create_t,wrapper_t> boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,T,TName,TPriority,TCtor>>::create_successful_impl<boost::di::v1_1_0::aux::true_type,example>(const boost::di::v1_1_0::aux::type<example> &) const' being compiled
1>        with
1>        [
1>            TConfig=config,
1>            TScope=boost::di::v1_1_0::scopes::deduce,
1>            TExpected=interface,
1>            T=implementation,
1>            TName=boost::di::v1_1_0::no_name,
1>            TPriority=void,
1>            TCtor=boost::di::v1_1_0::core::none
1>        ]
1>C:\Projects\di\example\pool_provider.cpp(119): message : see reference to function template instantiation 'T boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,implementation,TName,TPriority,TCtor>>::create<example,0>(void) const' being compiled
1>        with
1>        [
1>            T=example,
1>            TConfig=config,
1>            TScope=boost::di::v1_1_0::scopes::deduce,
1>            TExpected=interface,
1>            TName=boost::di::v1_1_0::no_name,
1>            TPriority=void,
1>            TCtor=boost::di::v1_1_0::core::none
1>        ]
1>C:\Projects\di\example\pool_provider.cpp(119): message : see reference to function template instantiation 'T boost::di::v1_1_0::core::injector<TConfig,boost::di::v1_1_0::core::pool<boost::di::v1_1_0::aux::type_list<>>,dependency,boost::di::v1_1_0::core::dependency<TScope,TExpected,implementation,TName,TPriority,TCtor>>::create<example,0>(void) const' being compiled
1>        with
1>        [
1>            T=example,
1>            TConfig=config,
1>            TScope=boost::di::v1_1_0::scopes::deduce,
1>            TExpected=interface,
1>            TName=boost::di::v1_1_0::no_name,
1>            TPriority=void,
1>            TCtor=boost::di::v1_1_0::core::none
1>        ]
```

File: include/boost/di.hpp [hunk 2]
```diff
@@ -43,6 +43,9 @@ BOOST_DI_CFG_FWD
   {}
 #elif defined(_MSC_VER)
 #define __MSVC__
+#if !defined(__has_extension)
+#define __has_extension(...) 1
```
Comment (kanstantsin-chernik) [inline include/boost/di.hpp:47]: `__has_extension` is not a thing in msvc. However it does support `__is_constructable`.
We can even employ `std::is_constructable` if defined

File: test/ut/type_traits/ctor_traits.cpp [hunk 3]
```diff
@@ -106,9 +106,9 @@ test ctors = [] {
   test_ctor_traits<ec, direct>();
   test_ctor_traits<conv, uniform>();
   test_ctor_traits<conv_explicit, direct, core::any_type_fwd<conv_explicit>, core::any_type_fwd<conv_explicit>>();
-  test_ctor_traits<vaarg, direct, core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>,
-                   core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>,
-                   core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>>();
+  //test_ctor_traits<vaarg, direct, core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>,
+  //                 core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>,
+  //                 core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>, core::any_type_fwd<vaarg>>();
```
Comment (kanstantsin-chernik) [inline test/ut/type_traits/ctor_traits.cpp:111]: These test broke. And, honestly, I have no freaking idea why. The error is below:
```
1>C:\Projects\di\test\ut\type_traits\ctor_traits.cpp(38,1): error C2338: fail
1>C:\Projects\di\test\ut\type_traits\ctor_traits.cpp(111): message : see reference to function template instantiation 'void type_traits::test_ctor_traits<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg,type_traits::direct,core::any_type_fwd<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg>,core::any_type_fwd<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg>,core::any_type_fwd<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg>,core::any_type_fwd<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg>,core::any_type_fwd<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg>,core::any_type_fwd<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg>,core::any_type_fwd<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg>,core::any_type_fwd<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg>,core::any_type_fwd<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg>,core::any_type_fwd<type_traits::<lambda_2f76908f4c1891ae561b2cb8bac97219>::()::vaarg>>(void)' being compiled
```

**Comment by kris-jusiak** [comment]
hmm :thinking: Let me check, will need to get the newest msvc to get the errors, thanks @kanstantsin-chernik for working on it :)

**Comment by kanstantsin-chernik** [comment]
This fix breaks VS 2015. We need to find something else

**Pull request closed without merging**